lib/vagrant-cucumber/step_definitions.rb

```diff
@@ -1,19 +1,43 @@
 require 'to_regexp'
 
 def push_snapshot(vmname)
-    vagrant_glue.vagrant_env.cli('snapshot', 'push', vmname)
+    case machine_provider(vmname)
+    when :libvirt
+        require 'sahara/session/factory'
+        ses = Sahara::Session::Factory.create(vagrant_glue.get_vm(vmname))
+        unless ses.is_snapshot_mode_on?
+            vagrant_glue.vagrant_env.cli('sandbox', 'on', vmname)
+            vagrant_glue.vagrant_env.cli('sandbox', 'commit', vmname)
+        end
+    when :no_machines
+        return
+    else
+        machine = vagrant_glue.get_vm(vmname)
+        vagrant_glue.vagrant_env.cli('snapshot', 'push', vmname) if machine.provider.capability(:snapshot_list).empty?
```

Comment:
Moving the `empty?` check here changes the behaviour of the `push_snapshot` function. Given that others may have written steps around this function (maybe to take an extra snapshot during the middle of a scenario) I believe this check has to live in the `there is a running VM called` step definition as before.